lib/block_view_lockups.go

```diff
@@ -0,0 +1,2178 @@
+	txMeta := txn.TxnMeta.(*CoinLockupMetadata)
+
+	// Check that the target profile public key is valid and that a profile corresponding to that public key exists.
+	var profileEntry *ProfileEntry
+	if len(txMeta.ProfilePublicKey) != btcec.PubKeyBytesLenCompressed {
+		return 0, 0, nil,
+			errors.Wrap(RuleErrorCoinLockupInvalidProfilePubKey, "_connectCoinLockup")
+	}
+	if !txMeta.ProfilePublicKey.IsZeroPublicKey() {
+		profileEntry = bav.GetProfileEntryForPublicKey(txMeta.ProfilePublicKey.ToBytes())
+		if profileEntry == nil || profileEntry.isDeleted {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinLockupOnNonExistentProfile, "_connectCoinLockup")
+		}
+	}
+
+	// Validate the lockup amount as non-zero. This is meant to prevent wasteful "no-op" transactions.
+	if txMeta.LockupAmountBaseUnits.IsZero() {
+		return 0, 0, nil,
+			errors.Wrap(RuleErrorCoinLockupOfAmountZero, "_connectCoinLockup")
+	}
+
+	// If this is a DeSo lockup, ensure the amount is less than 2**64 (maximum DeSo balance).
+	if txMeta.ProfilePublicKey.IsZeroPublicKey() && !txMeta.LockupAmountBaseUnits.IsUint64() {
+		return 0, 0, nil,
+			errors.Wrap(RuleErrorCoinLockupExcessiveDeSoLockup, "_connectCoinLockup")
+	}
+
+	// Validate the lockup expires in the future.
+	if txMeta.UnlockTimestampNanoSecs <= blockTimestamp {
+		return 0, 0, nil,
+			errors.Wrap(RuleErrorCoinLockupInvalidLockupDuration, "_connectCoinLockup")
+	}
+
+	// Compute the lockup duration in nanoseconds.
+	lockupDurationNanoSeconds := txMeta.UnlockTimestampNanoSecs - blockTimestamp
+
+	// Determine the hodler PKID to use.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(txn.PublicKey)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return 0, 0, nil,
+			errors.Wrap(RuleErrorCoinLockupInvalidTransactorPKID, "_connectCoinLockup")
+	}
+	hodlerPKID := transactorPKIDEntry.PKID
+
+	// Determine which profile PKID to use.
+	var profilePKID *PKID
+	if txMeta.ProfilePublicKey.IsZeroPublicKey() {
+		profilePKID = ZeroPKID.NewPKID()
+	} else {
+		profilePKIDEntry := bav.GetPKIDForPublicKey(txMeta.ProfilePublicKey.ToBytes())
+		if profilePKIDEntry == nil || profilePKIDEntry.isDeleted {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinLockupNonExistentProfile, "_connectCoinLockup")
+		}
+		profilePKID = profilePKIDEntry.PKID.NewPKID()
+	}
+
+	// Validate the transactor as having sufficient DAO Coin or DESO balance for the transaction.
+	var transactorBalanceNanos256 *uint256.Int
+	var prevTransactorBalanceEntry *BalanceEntry
+	var prevCoinEntry *CoinEntry
+	if profilePKID.IsZeroPKID() {
+		// NOTE: When spending balances, we need to check for immature block rewards. Since we don't have
+		//       the block rewards yet for the current block, we subtract one from the current block height
+		//        when spending balances.
+
+		// Check the DeSo balance of the user.
+		transactorBalanceNanos, err := bav.GetSpendableDeSoBalanceNanosForPublicKey(txn.PublicKey, blockHeight-1)
```

Comment:
Hmm... Do we do this in other places? I thought you don't need to subtract 1 for this to work but I can't remember. Just triple check that this is the same behavior as everywhere else, or that there's a good reason why it's different.

Reply:
Ugh I looked around and it seems like we call it with both blockHeight and blockHeight-1, and it's not clear which is actually the one we want. I think blockHeight (without the -1) is the more robust check because it will exclude block rewards from the tip. But I'm not 100% sure. @lazynina do you know if there's a good reason to call GetSpendableDeSoBalanceNanosForPublicKey with blockHeight-1 in UtxoView?